### ISSUE TITLE
Executables crash on startup with bundle_files<=2 on Python >=3.8

### ISSUE THREAD

**Issue by neodyne** [py2exe/py2exe#63]
os=Windows 10 x64, fully updated
python>=3.7, observed on version 3.8.6 (AMD64)
py2exe>=0.10.x (both releases that support 3.8, latest from pip or the first v0.10 from the Releases section)
in setup.py, bundle_files<=2
using only build-in python modules

For example, even [this sample program](https://github.com/py2exe/py2exe/issues/35#issuecomment-567361638), known to work with py2exe 0.9. The executable builds fine, without errors. Upon launching it crashes without any output, and creates a `Main.exe.dmp` in `%LOCALAPPDATA%\CrashDumps`.

Going back to previous versions, latest v0.9 works with Python 3.7, although with `bundle_files=0`, modules depending on _ssl (urllib in my case) throw an 'unspecified error', despite [a merged PR](https://github.com/py2exe/py2exe/pull/13).

Currently, only single file executables for my usecase that work are the ones generated by py2exe==0.9.3.2 on python==3.6 (and probably 3.5).

Please look into it when you'll have the chance.

**Comment by albertosottile** [comment]
Most likely, `bundle_files<=2` does not work well with versions 0.10.x. Unfortunately, I do not have the resources to run all the tests also with this option. I would encourage you to set `bundle_files=3` for your programs.

Nevertheless, if you could provide me a test script and a `setup.py` file in this issue, I could try to give another look at your specific problem when I have some spare time.

**Comment by neodyne** [comment]
Thank you for looking into this. Unfortunately, single executable is imperative for the project, for variety of reasons, so `bundle_files=3` is not viable for me.

As linked in my issue, for example a basic `Main.py`  from [this comment](https://github.com/py2exe/py2exe/issues/35#issuecomment-567361638),
```
f = open('mylog.log', 'w')
f.write('Hello world')
f.close()
```
with the `setup.py` from the same comment, or even a barebones one with just

```
from distutils.core import setup
import py2exe

setup(console=[{ "script": "Main.py"}],
    options={"py2exe": {
        "bundle_files": 0}}) # bundle_files<=2
```

 exhibits the described behavior (builds fine but crashes when run) with 0.10.x, on all currently supported python versions (3.6-3.8).

**Comment by albertosottile** [comment]
I managed to get some more information, but I am miles away from fixing this. I packed the example code with `bundle_files=2` and confirmed the crash. From Event Viewer, I found out that the exception code of the executable is `0xc0000005`, which means Access Violation. This, combined with the fault offset of `0x0` means that something was trying to access a `NULL` pointer. 

Running the bundled executable in an environment with `PY2EXE_VERBOSE` set to `2`, these are the last lines printed before the crash:
```
import '_memimporter' # <class '_frozen_importlib.BuiltinImporter'>
import zipextimporter # loaded from Zip C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\zipextimporter.pyc
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\ctypes\__init__.pyc
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\ctypes\__init__.pyc
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\_ctypes\__init__.pyc
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\_ctypes\__init__.py
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\_ctypes.pyc
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\_ctypes.py
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\_ctypes\__init__.pyc
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\_ctypes\__init__.py
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\_ctypes.pyc
# trying C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip\_ctypes.py
error loading _ctypes: can't find module '_ctypes'
# found _ctypes.pyd in zipfile C:\Users\Alberto\Documents\py2exe\tests\functional\bundlefiles_test\dist2\library.zip
```
which confirmed my suspicions: the executable runs fine as long as it has to rely on `zipimport`. The moment we need to import a PYD file from the zip, hence, to rely on `_memimporter`, the crash happens. 

After compiling the py2exe stubs with debug enabled, and after passing the py2exe and Python PDB debug symbols to WinDbg, I got the following trace of the crash:
```
 # Child-SP          RetAddr           Call Site
00 00000003`148fcab8 00007ff7`4390405d 0x0
01 00000003`148fcac0 00007ff7`439048b3 run_py3_8_win_amd64!MemoryLoadLibraryEx+0x22d [C:\projects\py2exe-iupeb\source\MemoryModule.c @ 450] 
02 00000003`148fcb00 00007ffe`093ad142 run_py3_8_win_amd64!FinalizeSections+0x113 [C:\projects\py2exe-iupeb\source\MemoryModule.c @ 174] 
[...]
```
which points to `locationDelta = (SIZE_T)(code - old_header->OptionalHeader.ImageBase);`, which is not really helpful. Furthermore, looking at symbols in the frame, the debugger reports 
```
00000003`148fcb10 unsigned int64 locationDelta = 1
``` 
which definitely does not help me. 

I will keep this comment as starting point to proceed further. One approach could be to define the `VERBOSE` constant in `MyLoadLibrary.c`, recompile and observe existing/insert further printouts. 

Another one could be to just upgrade the embedded copy of `MemoryModule` taking the latest version from https://github.com/fancycode/MemoryModule/tree/master. Unfortunately, the interface has changed a lot since the time this was embedded in py2exe and there are some new arguments needed that I am not sure how to provide.

In summary, this investigation confirmed that anything bundled with `bundle_files<=2` is indeed broken and will likely stay as it is for a while. If anyone has suggestions, ideas on how to proceed, or time to dig this further, comments and PRs are always welcome as usual.

**Comment by neodyne** [comment]
Update, sort of.

The bug might have been introduced in b25a41e28fd910db8c44dfc62e42bfa50195e900 with `pewalker` (and the `pefile` dependency), and persisted with the switch to `pescan` in 28bc95dee655d126b402b341b5e98811050377d0. After removing the `pescan`  [backup method](https://github.com/py2exe/py2exe/blob/master/py2exe/dllfinder.py#L132) from `dllfinder.py` in the latest master (= not using `pescan.py` at all) and building the py2exe wheel (MS Build Tools for VS2019, Python 3.7), the executables built with this py2exe version no longer crash.

Alas, if the wheel is built with python 3.8 with the same patch the resulting exe still crashes. So this may also be a compatibility issue with python 3.8+. Either way, the logic in pescan/pewalker, or the backup method likely warrants a closer look.

**Comment by albertosottile** [comment]
I doubt that `pefile` has anything to do with this, as it just packs different files in the bundle and this example already reduces the external dependencies to none: at the moment, we have troubles in importing standard library modules.

Indeed, though, the troubles seems to have started with 3.8, as the example runs fine with 3.7 and 3.6. Thanks for finding this. This is very helpful, both for debugging the issue but also as a workaround for all the users. I hope you do not mind that I pinned this issue and that I will change the title according to new findings.

**Comment by neodyne** [comment]
To avoid any misunderstanding:

if I install python <3.8, e.g. 3.7, and install py2exe==0.10.1.0 (current) with `pip install py2exe`, the crash is reproducible with the example in all cases. Only if i remove the `pescan` code from `dllfinder`, rebuild the wheel & install it to a fresh python 3.7 environment does it run without issues. I am not sure myself why it would matter but i went up [the commit history](https://github.com/py2exe/py2exe/compare/v0.9.3.2...v0.10.0.2) (skipping license/readme amendments and tests/appveyor fixes), on python 3.7, from py2exe 0.9.3.2, the last known good one. The breaking commit was b25a41e, adding `pewalker` (of which `pescan` is a later rewrite).

I think these may be 2 interconnected issues, on 3.8 it just plain doesn't work, but on 3.7 it broke.

**Comment by albertosottile** [comment]
Mmmh, then indeed we do have a misunderstanding. I performed the test described in this issue with **latest** py2exe (0.10.1.0) and  `bundle_files=2`. The test was a **PASS** on Python 3.6 and 3.7, and a fail on Python 3.8 and 3.9. 

I also just ran the same test in the CI flow on this repository, running on the latest development branch, and obtained exactly the same result: https://ci.appveyor.com/project/albertosottile/py2exe-iupeb/builds/37268270

In all the tests, all the code was not changed, including the `pefile` related parts.

So, to clarify, I **can** run `bundle_file=2` with the existing, untouched codebase on Python <=3.7. Therefore, I would encourage you to try again in an isolated environment (e.g. Docker) or to double check your environment. I would also encourage other users that need this option to use Python 3.7 or lower, hence the new issue title.

Now, your last edit seems to agree with these findings. About `bundle_files=0`, for the moment, I would address one issue at the time, which is `bundle_files=2`, because we would need this for any lower bundle value. If we ever manage to fix this, then we can try to reduce this further down to the 1 and 0 options.

**Comment by albertosottile** [comment]
Ok. after I played a while with WinDbg I identified the problem, which was actually not in `MemoryModule.c` but in `python-dynload.c` (I am not sure why the first time the debugger pointed me there, probably it lacked some symbols). Anyway, I now built preliminary wheels that include this fix and also added the same test down to `bundle_files=0`, and they all passed: https://ci.appveyor.com/project/albertosottile/py2exe-iupeb/builds/37274012

Just to reiterate, I do not plan to support complex programs and all the package imports when `bundle_files<3` but, for simple programs that depend on standard library and not much more, it should work. 

@neodyne I would really appreciate if you could download the preliminary wheels from https://bintray.com/py2exe/py2exe/py2exe/0.10.2.0#files, install them, and test them again on your system. Please let me know if they work also for you, as I plan to include these changes in the next release of py2exe. Thanks again for having reported this and for your time.

**Comment by neodyne** [comment]
I can confirm executables packed with the new py2exe wheels do not crash. The problem with the hook for `_ssl` briefly mentioned in the original report remains, but I shall open a separate issue for that. Feel free to close this one as solved.

Thank you for the very quick fix, for your patience and in general, keeping py2exe on python 3 alive and kicking. It is greatly appreciated.

**Comment by albertosottile** [comment]
Fixed in [0.10.2.0](https://github.com/py2exe/py2exe/releases/tag/v0.10.2.0).
